Move MemoryFormat/Layout to headeronly

mikaylagawarecki opened the pull request:
~This PR does change the semantics of the >> operator by using STD_TORCH_CHECK to throw the error instead of TORCH_CHECK. Jane (who is writing this message) thinks it is okay because it is the error case when an invalid MemoryFormat or Layout is getting passed into >>, so the UX benefits of TORCH_CHECK over STD_TORCH_CHECK there are not significant enough to warrant making a new copy of Layout and MemoryFormat's >> APIs.~

Never mind! We shouldn't change TORCH_CHECK to STD_TORCH_CHECK for core usage ever, cuz the traceback info and c10::Error is very much desired!! So the solution is to not migrate the >>s. I pushed new commits to the stack to remove the >> code, but for reference, https://github.com/pytorch/pytorch/pull/168034/commits/8a30179fab3a52ef23fffe19ddad765d5a230ca5 has all the code that I ended up deleting.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #168034
* #167962
* #167804
* #167803
* #167802
* #168025

mikaylagawarecki committed:
Move MemoryFormat/Layout to headeronly

[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Move MemoryFormat/Layout to headeronly"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Move MemoryFormat/Layout to headeronly"




[ghstack-poisoned]

janeyx99 committed:
Update on "Move MemoryFormat/Layout to headeronly AND use STD_TORCH_CHECK"




[ghstack-poisoned]

janeyx99 commented on torch/headeronly/core/Layout.h:
```diff
@@ -0,0 +1,70 @@
+    default:
+      // Note: We can't use TORCH_CHECK here as it's not header-only
+      // Callers should ensure valid layout values
+      return stream << "Unknown";
```
Actually....since that changes semantics...maybe an alternative is to just not migrate the >> functions

janeyx99 committed:
Update on "Move MemoryFormat/Layout to headeronly"

~This PR does change the semantics of the >> operator by using STD_TORCH_CHECK to throw the error instead of TORCH_CHECK. Jane (who is writing this message) thinks it is okay because it is the error case when an invalid MemoryFormat or Layout is getting passed into >>, so the UX benefits of TORCH_CHECK over STD_TORCH_CHECK there are not significant enough to warrant making a new copy of Layout and MemoryFormat's >> APIs.~

Never mind! We shouldn't change TORCH_CHECK to STD_TORCH_CHECK for core usage ever, cuz the traceback info and c10::Error is very much desired!! So the solution is to not migrate the >>s. I pushed new commits to the stack to remove the >> code, but for reference, https://github.com/pytorch/pytorch/pull/168034/commits/8a30179fab3a52ef23fffe19ddad765d5a230ca5 has all the code that I ended up deleting.




[ghstack-poisoned]

janeyx99 commented:
@pytorchbot merge
